include/common/mem.h

```diff
@@ -0,0 +1,58 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+#ifndef __LABWC_MEM_H
+#define __LABWC_MEM_H
+
+#include <stdlib.h>
+
+/*
+ * As defined in busybox, weston, etc.
+ * Allocates zero-filled memory; calls exit() on error.
+ * Returns NULL only if (size == 0).
+ */
+void *xzalloc(size_t size);
+
+/* Type-safe macros in the style of C++ new/new[] */
+#define znew(type)       ((type *)xzalloc(sizeof(type)))
+#define znew_n(type, n)  ((type *)xzalloc((n) * sizeof(type)))
+
+/*
+ * Variant of znew() that infers type from the passed pointer.
+ * Preferred over plain znew() in pointer declarations since it
+ * does not require repeating the type name.
+ *
+ * Example:
+ *   struct wlr_box *box = znew_for(box);
+ *
+ * Note:
+ *   In C23 "typeof" will be standard. Until then, __typeof__ works
+ *   with both GCC and clang (do we care about other compilers?)
+ */
```

Comment:
`__typeof__` is already used in `wl_list_for_each` and `FOR_EACH()` in `ssd.h` so its usage is fine IMO.

With the added type check we could go back to `znew_for(*ptr)` though, couldn't we?
E.g.
```c
define znew_for(ptr)  ((__typeof__(&ptr))xzalloc(sizeof(ptr)))
```

Did the conversion for that here to get a feel of the API: https://github.com/Consolatis/labwc/tree/wip/znew_for

Reply:
If we could unify the two safely with `__typeof__`, that would be very slick.